GDB-11318 implement global notification component

plamen-yordanov opened the pull request:
## What
Implement global notification component

## Why
To be able to display statuses of different ongoing operations

## How
- Created `onto-operations-notification` dropdown button, which displays different operations as icons and respective count. The button is a dropdown toggle, so when clicked, it opens a list of different operations. When an individual operation is clicked, it redirects to a page, related to the operation

- Refactored `onto-header` to display/hide the new notification component and poll the operations on every 2 seconds

- Migrated `fibonacci-generator`, which is currently used to slow down requests, if errors occur

- Added models and services

## Testing
jest and cypress

## Screenshots
![Screenshot from 2025-03-12 13-49-16](https://github.com/user-attachments/assets/bbbbf89c-b2f0-43b1-87b8-15b5c3919f26)

![Screenshot from 2025-03-12 13-52-08](https://github.com/user-attachments/assets/897bc64d-4eef-4973-8819-b14da35d623e)

## Checklist
- [x] Branch name
- [x] Target branch
- [x] Commit messages
- [x] Squash commits
- [x] MR name
- [x] MR Description
- [x] Tests

svilenvelikov commented on packages/shared-components/src/components/onto-operations-notification/onto-operations-notification.scss:
```diff
@@ -0,0 +1,165 @@
+    border-radius: 0;
+    list-style: none;
+    padding: 0;
+    -moz-transition: all 150ms ease-out; /* FF4+ */
```
we should add the autoprefixer library and write just the standard's compliant css rules

plamen-yordanov replied:
In a different task, I assume?

svilenvelikov replied:
yes

svilenvelikov commented on packages/shared-components/src/components/onto-operations-notification/onto-operations-notification.tsx:
```diff
@@ -0,0 +1,142 @@
+
+
+  private toOperationsGroupSummary() {
+    const groupToOperationSummary = new Map<OperationGroup, OperationGroupSummary>();
```
Is it possible this mapping building to be moved in the activeOperations model maybe. Better avoid exposing the model internals in components to reduce the coupling

plamen-yordanov replied:
moved inside the model

plamen-yordanov commented on packages/api/src/models/monitoring/operation-group-summary.ts:
```diff
@@ -0,0 +1,21 @@
+
+  constructor(data: OperationGroupSummary) {
+    super();
+    this.id = window.crypto.randomUUID();
```
Can the service/util use `window.crypto.randomUUID()` internally, or should I implement something?

svilenvelikov replied:
Yes. Just beware that our services are singletons and we don't want to cache the window inside.

plamen-yordanov replied:
moved to `generator-utils`

plamen-yordanov commented on packages/api/src/models/monitoring/operation-status-summary.ts:
```diff
@@ -0,0 +1,22 @@
+  constructor(operationStatusSummary: Partial<OperationStatusSummary>) {
+    super();
+    this.status = operationStatusSummary.status || OperationStatus.INFORMATION;
+    this.allRunningOperations = MapperProvider.get(OperationListMapper).mapToModel(operationStatusSummary.allRunningOperations || []);
```
This won't default to a native array. It will pass a native empty array to the mapper, which will result in an empty `OperationList`

svilenvelikov replied:
I see. My bad.

plamen-yordanov commented on packages/shared-components/src/components/onto-header/onto-header.tsx:
```diff
@@ -43,11 +77,56 @@ export class OntoHeader {
-  disconnectedCallback(): void {
-    this.subscriptions.unsubscribeAll();
+  private startOperationPolling() {
+    this.pollingInterval = setInterval(() => {
```
The legacy workbench currently starts polling, when the component is loaded, without clearing the interval beforehand. The interval is cleared upon `$destroy`.

![image](https://github.com/user-attachments/assets/3a40082c-572e-4188-9371-b7a86e6a08ed)

The current implementation will stop polling, if the repositoryContext sends `repositoryId: undefined` as well. Otherwise it will start and stop, like the legacy

plamen-yordanov replied:
added `clearInterval` at the start just in case

plamen-yordanov commented on packages/shared-components/src/components/onto-header/onto-header.tsx:
```diff
@@ -9,30 +22,51 @@ import { ServiceProvider, LicenseContextService, License, SubscriptionList } fro
+  private readonly fibonacciGenerator = new FibonacciGenerator();
+
+  private repositoryId?: string;
+  private pollingInterval: NodeJS.Timeout;
```
added `number` type. Also added `window.setInterval` to ensure typescript knows it's a browser interval and not a nodejs. If you don't like depending directly on window here, the other option is //@ts-ignore, since typescript doesn't know, which interval it is.

plamen-yordanov commented:
Jenkins retry build

svilenvelikov commented on packages/api/src/models/monitoring/operation-status-summary.ts:
```diff
@@ -19,4 +23,35 @@ export class OperationStatusSummary extends Model<OperationStatusSummary> {
+    });
+
+    const operationGroupArray = Array.from(groupToOperationSummary.values());
+    return MapperProvider.get(OperationGroupSummaryListMapper).mapToModel(operationGroupArray);
```
OK, but now you put a service dependency in the model. Although it is a mapper, I don't think it's a good idea. You can return the array from this method and mode this mapping in the client component.

plamen-yordanov replied:
Should I limit mapper usage only to constructors? I've used list mappers in other models and I'm not sure, where to draw the line

svilenvelikov replied:
Models should not have dependencies than other models. If you've used mappers in models, then those should also be refactored.

plamen-yordanov replied:
Removed the dependency entirely. Moved the `OperationGroupSummaryListMapper#mapToModel` inside `OperationSummaryMapper`

plamen-yordanov commented:
Also: 
- Replaced `onto-operations-monitoring` with `onto-header` in the page file to use the same response mock and model mapping.
- Changed the component dropdown to close upon link click, since this is the original behavior
